pandas/core/groupby/groupby.py

```diff
@@ -588,6 +588,8 @@ class BaseGroupBy(PandasObject, SelectionMixin[NDFrameT], GroupByIndexingMixin):
         "obj",
         "observed",
         "sort",
+        "observed_grouper",
+        "observed_exclusions",
```

Comment:
Instead of this, I recommend adding it as a cached method on the `BaseGrouper` class in `ops.py`.

```
@cache_readonly
def observed_grouper(self):
    if all(ping._observed for ping in self.groupings):
        return self
    grouper = BaseGrouper(...)
    return grouper
```

For this to work, you also need to do the same to Grouping:

```
@cache_readonly
def observed_grouping(self):
    if self._observed:
        return self
    grouping = Grouping(...)
    return grouping
```

and use the observed_groupings in the BaseGrouper call above. For BinGrouper, I think you can just always return `self` (doesn't change behavior on to observed=True/False).

Also, you can ignore `exclusions` - this is independent of the grouping data stored in BaseGrouper/Grouping.

Reply:
Ah my bad, I have made the changes as suggested